### PR TITLE
[SYCLomatic] make the option "--enable-profiling" is enabled on demand

### PR DESCRIPTION
**Pull request by tomflinda**
Signed-off-by: chenwei.sun <chenwei.sun@intel.com>

### CODE REVIEW

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 1]
```diff
@@ -181,7 +181,7 @@ DPCT_NON_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::opt<bool, true>), CheckUnico
            llvm::cl::cat(DPCTCat), llvm::cl::location(CheckUnicodeSecurityFlag))
 
 DPCT_NON_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::opt<bool, true>), EnablepProfiling, "enable-profiling",
-           llvm::cl::desc("Enable SYCL queue profiling in helper functions. Default: off.\n"),
+           llvm::cl::desc("Enable SYCL queue profiling in helper functions on demand. \n"),
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc]: default: auto  (auto: mean off: if tool conclude profiling is not required, on: if tool conclude profiling is required).

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 2]
```diff
@@ -8106,6 +8106,44 @@ void EventAPICallRule::handleOrdinaryCalls(const CallExpr *Call) {
 
 REGISTER_RULE(EventAPICallRule, PassKind::PK_Migration)
 
+void ProfilingEnableOnDemandRule::registerMatcher(MatchFinder &MF) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp:8109]: Can it be merged into EventAPICallRule?

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 3]
```diff
@@ -870,8 +870,12 @@ void DpctFileInfo::insertHeader(HeaderType Type, unsigned Offset,
...
-      OS << "#define DPCT_PROFILING_ENABLED" << getNL();
+
+    // Add the label for profiling macro "DPCT_PROFILING_ENABLED", which will be
+    // replaced by "#define DPCT_PROFILING_ENABLED" or not in the post
+    // replacement.
+    OS << "{{NEEDREPLACEP0}}";
+
     if (DpctGlobalInfo::getUsmLevel() == UsmLevel::UL_None)
       OS << "#define DPCT_USM_LEVEL_NONE" << getNL();
     concatHeader(OS, getHeaderSpelling(Type));
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp:880]: Line 907 also has getEnablepProfilingFlag

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 4]
```diff
@@ -870,8 +870,12 @@ void DpctFileInfo::insertHeader(HeaderType Type, unsigned Offset,
     return insertHeader(OS.str(), FirstIncludeOffset,
                         InsertPosition::IP_AlwaysLeft);
   case HT_SYCL:
-    if (DpctGlobalInfo::getEnablepProfilingFlag())
-      OS << "#define DPCT_PROFILING_ENABLED" << getNL();
+
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]:
```suggestion
```

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 5]
```diff
@@ -8106,6 +8106,44 @@ void EventAPICallRule::handleOrdinaryCalls(const CallExpr *Call) {
...
+
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]:
```suggestion
```

**Review by zhiweij1 — CHANGES_REQUESTED** [review]
Please update the PR

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 5 above)
Comment (tomflinda) [reply clang/lib/DPCT/ASTTraversal.cpp]: okay

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 4 above)
Comment (tomflinda) [reply clang/lib/DPCT/AnalysisInfo.cpp]: okay

**Review by zhiweij1 — APPROVED** [review]

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 6]
```diff
@@ -180,9 +180,13 @@ DPCT_NON_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::opt<bool, true>), CheckUnico
...
            llvm::cl::cat(DPCTCat), llvm::cl::location(CheckUnicodeSecurityFlag))
 
-DPCT_NON_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::opt<bool, true>), EnablepProfiling, "enable-profiling",
-           llvm::cl::desc("Enable SYCL queue profiling in helper functions. Default: off.\n"),
-           llvm::cl::cat(DPCTCat), llvm::cl::location(EnablepProfilingFlag))
+DPCT_NON_ENUM_OPTION(
+    DPCT_OPT_TYPE(static llvm::cl::opt<bool, true>), EnablepProfiling,
+    "enable-profiling",
+    llvm::cl::desc("Enable SYCL queue profiling in helper functions. default: "
+                   "auto\n(auto: mean off: if tool concludes profiling is not "
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc]:
```suggestion
                   "auto (enable-profiling if \n the tool deduces that profiling is required during migration, else not).
```

Comment (tomflinda) [reply clang/include/clang/DPCT/DPCTOptions.inc]: Updated.

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**